mmv1/third_party/terraform/tests/resource_compute_security_policy_test.go.erb

```diff
@@ -274,6 +274,30 @@ func TestAccComputeSecurityPolicy_withRateLimitWithRedirectOptions(t *testing.T)
 	})
 }
 
+<% unless version == 'ga' -%>
+func TestAccComputeSecurityPolicy_withRateLimit_withEnforceOnKeyConfigs(t *testing.T) {
```

Comment:
The test failed. You can check the errr message by clicking "Error Message" next to the failed test.

Reply:
Hi there! 

The test failed because we need to change the enforceOnKey, this provider has "default" value and automatically is inserted on the tests:

![image](https://user-images.githubusercontent.com/120479960/213205331-17807c7e-4528-4893-b9fb-f9b889381d5d.png)


The enforceOnKeyConfigs is a new providers that must not has the enforceOnKey in the same script.

Check the image from API:
![image](https://user-images.githubusercontent.com/120479960/213204230-4f78b376-dad9-4ed0-9408-c3d61e855648.png)

Check the API doc bellow if you wish:
https://cloud.google.com/compute/docs/reference/rest/beta/securityPolicies

We can't remove the default value from enforceOnKey, we are search how we can solved this. When we run the test with enforceOnKey default value on we receive this error:

![image](https://user-images.githubusercontent.com/120479960/213204328-7258cf03-570a-403a-a02d-c5d0d2a963b3.png)

When we run the test without  enforceOnKey default value the test passes.

Reply:
I tested after removing the Default: "ALL" for the field `enforce_on_key` in this file and then run the test locally.

Request: 
```
{
 "description": "updated description",
 "name": "tf-test-418rgij846",
 "rules": [
  {
   "action": "throttle",
   "description": "default rule",
   "headerAction": {},
   "match": {
    "config": {
     "srcIpRanges": [
      "*"
     ]
    },
    "versionedExpr": "SRC_IPS_V1"
   },
   "preview": false,
   "priority": 2147483647,
   "rateLimitOptions": {
    "conformAction": "allow",
    "exceedAction": "deny(403)",
    "rateLimitThreshold": {
     "count": 100,
     "intervalSec": 60
    }
   }
  }
 ]
}
```
     

Response:

```
{
 "kind": "compute#securityPolicy",
 "id": "184350908738827980",
 "creationTimestamp": "2023-01-18T11:27:31.714-08:00",
 "name": "tf-test-418rgij846",
 "description": "updated description",
 "rules": [
  {
   "kind": "compute#securityPolicyRule",
   "description": "default rule",
   "priority": 2147483647,
   "match": {
    "versionedExpr": "SRC_IPS_V1",
    "config": {
     "srcIpRanges": [
      "*"
     ]
    }
   },
   "action": "throttle",
   "preview": false,
   "rateLimitOptions": {
    "rateLimitThreshold": {
     "count": 100,
     "intervalSec": 60
    },
    "conformAction": "allow",
    "exceedAction": "deny(403)"
   },
   "headerAction": {}
  }
 ],
 "fingerprint": "mQvRGsyan_U=",
 "selfLink": "https://www.googleapis.com/compute/v1/projects/local-mediator-361721/global/securityPolicies/tf-test-418rgij846",
 "type": "CLOUD_ARMOR"
}

```

In the response, I do not see the field `enforceOnKey`. Does that mean `enforceOnKey` has the default value ALL for this rule?

Reply:
Is the response expected from API side?

Reply:
Hello there! 

About your test: yes, you're correct, you locally removed the Default: “ALL” from the  `enforceOnKey` field and the response doesn't show the value, right? This is expected to happen when you make this change.

The default value when inserted in a provider is automatically inserted when you generate a terraform test or a terraform script, that's the purpose of using the Default field.

I looked better about the  in [terraform-plugin-sdk](https://github.com/hashicorp/terraform-plugin-sdk) (the library used to build providers) and the description about what it does when: the attribute Default

	// Default indicates a value to set if this attribute is not set in the
	// configuration. Default cannot be used with DefaultFunc or Required.
	// Default is only supported if the Type is TypeBool, TypeFloat, TypeInt,
	// or TypeString. Default cannot be used if the Schema is directly an
	// implementation of an Elem field of another Schema, such as trying to
	// set a default value for a TypeList or TypeSet.
	//
	// Changing either Default can be a breaking change, especially if the
	// attribute has ForceNew enabled. If a default needs to change to align
	// with changing assumptions in an upstream API, then it may be necessary
	// to also implement resource state upgrade functionality to change the
	// state to match or update read operation logic to align with the new
	// default.
	Default interface{}



[Link to the code above](https://github.com/hashicorp/terraform-plugin-sdk/blob/6c49b02251cc161b7e4ad8f80b1d5e856364335f/helper/schema/schema.go#L160-L173).

The `enforceOnKeyConfigs` is a new attribute and conflicts with `enforceOnKey`, so only one must be set at a time, as stated in the [Google API documentation (https://cloud.google.com/compute/docs/reference/rest/beta/securityPolicies#:~:text=If%20enforceOnKeyConfigs%20is%20specified%2C%20enforceOnKey%20must%20not%20be%20specified): 

![image](https://user-images.githubusercontent.com/120479960/213748943-139d5ba2-ffc5-468b-976d-d44ec27868ef.png)


I double checked and have created a test with `enforceOnKeyConfigs` fields,and tested locally with `enforceOnKey` Default: “ALL” inserted and I got an error with the message: “Only one of enforceOnKey and enforceOnKeyConfigs can be specified”. When I removed the Default: “ALL” from `enforceOnKey` and ran the test again, it passed successfully.

![image](https://user-images.githubusercontent.com/120479960/213749112-4a4e088f-2e4f-4e1d-ad9f-738f9f7de7f7.png)
_(Conflict error between `enforceOnKey` and `enforceOnKeyConfigs` when default is inserted)_

You can check on SDK documentation about the enforceOnKeyConfigs [here](https://pkg.go.dev/google.golang.org/api/compute/v0.beta#SecurityPolicyRuleRateLimitOptions.EnforceOnKeyConfigs)

I tried to remove the `enforceOnKey` Default: “ALL” field, but during the PR I got the **breaking change**, I can’t remove this.

![image](https://user-images.githubusercontent.com/120479960/213749226-5d04f109-00a6-4938-ba3e-868bed63565f.png)

So, our understanding is that the magic module pipeline runs some additional tests, and for some reason it doesn’t handle this conflicting case.

Currently, there is only one test that tests `enforceOnKey`.

To test the enforceOnKeyConfigs, we cannot use enforceOnKey, but it sounds like a magic-module pipeline does some additional tests that automatically insert an enforceOnKey, enforcing that when we run locally the test, it PASS with success.


We saw 2 possible paths:

Create the tests in different steps, that is, keep the current test that only tests the `Default: “ALL”`. And add another test that does the `enforceOnKeyConfigs` tests.
Keep the current test and do not create the `enforceOnKeyConfigs` tests, following how the other `enforceOnKeyConfigs` were implemented.

We believe that the first option would be the better way, but from our understanding, it would still have the problem (About the conflict between  `enforceOnKey` e `enforceOnKeyConfigs`) .

We would like to know your suggestion about this way or other ways to solve this.

Reply:
1. Because of the default value `ALL` for the field `rule.rate_limit_options.enforce_on_key`, when the field `enforceOnKeyConfigs` is set, an error is returned from service side: “Only one of enforceOnKey and enforceOnKeyConfigs can be specified”. That is the reason to remove the default value "ALL" from Terraform provider. My question is that without default value "ALL" for the field `rule.rate_limit_options.enforce_on_key`, and the user does not set the field `enforceOnKeyConfigs`, will the rule work the same as before the default value "ALL" is removed from service side? Can you please double check it with the service team?


2. The  removal of the default value of the field `rule.rate_limit_options.enforce_on_key` is a breaking change and this change needs to wait for the next major release.  I tested it locally and terraform plan is not empty with the local provider version. The current release is hashicorp/google v4.50.0 and the next major release is v5.0.0. But the major release needs to be planned at least 6 month in advance and we do not have the plan yet. https://g3doc.corp.google.com/company/teams/terraform/users/getting-started/releases.md?cl=head#major-releases

Added @ScottSuarez for the advice how to handle this case.

Reply:
Hey !! Sorry for the delay. 

My advice here would be 
1. to remove the default field from the resource and add Computed to the field. 
2. @zli82016 , try to test this scenario to make sure that there are no negative side effects. The one I am aware of will be that the only way to /unset/ `enforce_on_key` will be by setting it to an empty string.. Adding this to the documentation would be nice to have.
3. Create a followup issue to remove this computed field from this property. Label this issue with breaking change so we can triage it in the 5.0 release.

Reply:
Thanks, @ScottSuarez . I removed the default value for the field and added Computed to the field. After applying the new provider built locally, terraform plan is not empty.

Reply:
Config: https://paste.googleplex.com/4578722240790528 
Terraform plan output:  https://paste.googleplex.com/6415383668981760

Reply:
@afpesseti , after talked with the team, we are going to do the following:

1. Add `""` to the list of allowed values for the field enforceOnKey to allow the user to specify `""` to avoid conflict with the new field. https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb#L265
2. In the tests, when testing with the new field enforceOnKeyConfigs, in the config add  
```
enforce_on_key = ""
```
to avoid conflict.

4. In the doc of the new field enforceOnKeyConfigs, add a note that to set the field, enforce_on_key needs to be empty string.

Reply:
Hello there! Thank you both for the help! About the new steps, just so there are no loose ends: are they going to be done by your end? Should we also do any of the listed steps on our end?
Just double checking! :D

Reply:
Hello, @afpesseti , can you please do the listed steps? If you need any help, feel free to let us know. Thanks.

Reply:
Hi There! Steps done :D Thank you so much for help us ❤️